File: test/language/explicit_parameters/cpp/ExplicitBitmaskParamTest.cpp
#include <cstdio>
#include <vector>
#include <string>

#include "gtest/gtest.h"

#include "explicit_parameters/ExplicitParametersDb.h"

#include "zserio/RebindAlloc.h"
#include "zserio/StringConvertUtil.h"

namespace explicit_parameters
{
namespace explicit_bitmask_param
{

using allocator_type = ExplicitParametersDb::allocator_type;
using string_type = zserio::string<zserio::RebindAlloc<allocator_type, char>>;
template <typename T>
using vector_type = std::vector<T, zserio::RebindAlloc<allocator_type, T>>;

class ExplicitBitmaskParamTest : public ::testing::Test
{
public:
    ExplicitBitmaskParamTest()
    {
        std::remove(DB_FILE_NAME);

        m_database = new ExplicitParametersDb(DB_FILE_NAME);
        m_database->createSchema();
    }

    ~ExplicitBitmaskParamTest()
    {
        delete m_database;
    }

protected:
    void fillBitmaskParamTableRow(BitmaskParamTable::Row& row, uint32_t id, const string_type& name)
    {
        row.setId(id);
        row.setName(name);

        TestBlob testBlob1;
        vector_type<uint8_t>& values1 = testBlob1.getValues();
        for (uint32_t i = 0; i < BITMASK_PARAM_TABLE_COUNT1.getValue(); ++i)
            values1.push_back(static_cast<uint8_t>(id));
        row.setBlob1(testBlob1);

        TestBlob testBlob2;
        vector_type<uint8_t>& values2 = testBlob2.getValues();
        for (uint32_t i = 0; i < BITMASK_PARAM_TABLE_COUNT2.getValue(); ++i)
            values2.push_back(static_cast<uint8_t>(id + 1));
        row.setBlob2(testBlob2);

        TestBlob testBlob3;
        vector_type<uint8_t>& values3 = testBlob3.getValues();
        for (uint32_t i = 0; i < BITMASK_PARAM_TABLE_COUNT1.getValue(); ++i)
            values3.push_back(static_cast<uint8_t>(id + 2));
        row.setBlob3(testBlob3);
    }

    void fillBitmaskParamTableRows(vector_type<BitmaskParamTable::Row>& rows)
    {
        rows.clear();
        for (uint32_t id = 0; id < NUM_BITMASK_PARAM_TABLE_ROWS; ++id)
        {
            const string_type name = "Name" + zserio::toString<allocator_type>(id);
            BitmaskParamTable::Row row;
            fillBitmaskParamTableRow(row, id, name);
            rows.push_back(row);
        }
    }

    static void checkBitmaskParamTableRow(const BitmaskParamTable::Row& row1,
            const BitmaskParamTable::Row& row2)
    {
        ASSERT_EQ(row1.getId(), row2.getId());
        ASSERT_EQ(row1.getName(), row2.getName());
        ASSERT_EQ(row1.getBlob1(), row2.getBlob1());
        ASSERT_EQ(row1.getBlob2(), row2.getBlob2());
        ASSERT_EQ(row1.getBlob3(), row2.getBlob3());

        // check reused explicit count parameter
        ASSERT_EQ(row2.getBlob1().getCount(), row2.getBlob3().getCount());
    }

    static void checkBitmaskParamTableRows(const vector_type<BitmaskParamTable::Row>& rows1,
            const vector_type<BitmaskParamTable::Row>& rows2)
    {
        ASSERT_EQ(rows1.size(), rows2.size());
        for (size_t i = 0; i < rows1.size(); ++i)
            checkBitmaskParamTableRow(rows1[i], rows2[i]);
    }

    class BitmaskParamTableParameterProvider : public BitmaskParamTable::IParameterProvider
    {
    public:
        virtual TestBitmask getCount1(BitmaskParamTable::Row&)
        {
            return BITMASK_PARAM_TABLE_COUNT1;
        }

        virtual TestBitmask getCount2(BitmaskParamTable::Row&)
        {
            return BITMASK_PARAM_TABLE_COUNT2;
        }
    };

    ExplicitParametersDb* m_database;

    static const char DB_FILE_NAME[];

    static const uint32_t NUM_BITMASK_PARAM_TABLE_ROWS;
    static const TestBitmask BITMASK_PARAM_TABLE_COUNT1;
    static const TestBitmask BITMASK_PARAM_TABLE_COUNT2;
};

const char ExplicitBitmaskParamTest::DB_FILE_NAME[] =
        "language/explicit_parameters/explicit_bitmask_param_test.sqlite";

const uint32_t ExplicitBitmaskParamTest::NUM_BITMASK_PARAM_TABLE_ROWS = 5;
const TestBitmask ExplicitBitmaskParamTest::BITMASK_PARAM_TABLE_COUNT1 = TestBitmask::Values::TEN;
const TestBitmask ExplicitBitmaskParamTest::BITMASK_PARAM_TABLE_COUNT2 = TestBitmask::Values::ELEVEN;

TEST_F(ExplicitBitmaskParamTest, readWithoutCondition)
{
    BitmaskParamTable& bitmaskParamTable = m_database->getBitmaskParamTable();

    BitmaskParamTableParameterProvider parameterProvider;
    vector_type<BitmaskParamTable::Row> writtenRows;
    fillBitmaskParamTableRows(writtenRows);
    bitmaskParamTable.write(parameterProvider, writtenRows);

    BitmaskParamTable::Reader reader = bitmaskParamTable.createReader(parameterProvider);

    vector_type<BitmaskParamTable::Row> readRows;
    while (reader.hasNext())
        readRows.push_back(reader.next());
    checkBitmaskParamTableRows(writtenRows, readRows);
}

TEST_F(ExplicitBitmaskParamTest, readWithCondition)
{
    BitmaskParamTable& bitmaskParamTable = m_database->getBitmaskParamTable();

    BitmaskParamTableParameterProvider parameterProvider;
    vector_type<BitmaskParamTable::Row> writtenRows;
    fillBitmaskParamTableRows(writtenRows);
    bitmaskParamTable.write(parameterProvider, writtenRows);

    const string_type condition = "name='Name1'";
    BitmaskParamTable::Reader reader = bitmaskParamTable.createReader(parameterProvider, condition);

    ASSERT_TRUE(reader.hasNext());
    BitmaskParamTable::Row readRow = reader.next();
    ASSERT_FALSE(reader.hasNext());

    const size_t expectedRowNum = 1;
    checkBitmaskParamTableRow(writtenRows[expectedRowNum], readRow);
}

TEST_F(ExplicitBitmaskParamTest, update)
{
    BitmaskParamTable& bitmaskParamTable = m_database->getBitmaskParamTable();

    BitmaskParamTableParameterProvider parameterProvider;
    vector_type<BitmaskParamTable::Row> writtenRows;
    fillBitmaskParamTableRows(writtenRows);
    bitmaskParamTable.write(parameterProvider, writtenRows);

    const uint64_t updateRowId = 3;
    BitmaskParamTable::Row updateRow;
    fillBitmaskParamTableRow(updateRow, updateRowId, "UpdatedName");
    const string_type updateCondition = "id=" + zserio::toString<allocator_type>(updateRowId);
    bitmaskParamTable.update(parameterProvider, updateRow, updateCondition);

    BitmaskParamTable::Reader reader = bitmaskParamTable.createReader(parameterProvider, updateCondition);

    ASSERT_TRUE(reader.hasNext());
    BitmaskParamTable::Row readRow = reader.next();
    ASSERT_FALSE(reader.hasNext());
    checkBitmaskParamTableRow(updateRow, readRow);
}

} // namespace explicit_bitmask_param
} // namespace explicit_parameters

File: test/language/explicit_parameters/python/ExplicitBitmaskParamTest.py
import unittest
import os

from testutils import getZserioApi, getApiDir

class ExplicitBitmaskParamTest(unittest.TestCase):
    @classmethod
    def setUpClass(cls):
        cls.api = getZserioApi(__file__, "explicit_parameters.zs")
        cls._fileName = os.path.join(getApiDir(os.path.dirname(__file__)), "explicit_bitmask_param_test.sqlite")

        test_api = cls.api.explicit_bitmask_param
        class BitmaskParamTableParameterProvider(test_api.BitmaskParamTable.IParameterProvider):
            @staticmethod
            def count1(_row):
                return test_api.TestBitmask.Values.TEN

            @staticmethod
            def count2(_row):
                return test_api.TestBitmask.Values.ELEVEN

        cls.BitmaskParamTableParameterProvider = BitmaskParamTableParameterProvider

    def setUp(self):
        if os.path.exists(self._fileName):
            os.remove(self._fileName)
        self._database = self.api.ExplicitParametersDb.from_file(self._fileName)
        self._database.create_schema()

        self._bitmaskParamTableCount1 = self.api.explicit_bitmask_param.TestBitmask.Values.TEN
        self._bitmaskParamTableCount2 = self.api.explicit_bitmask_param.TestBitmask.Values.ELEVEN

    def tearDown(self):
        self._database.close()

    def testDeleteTable(self):
        self.assertTrue(self._isTableInDb())

        testTable = self._database.bitmask_param_table
        testTable.delete_table()
        self.assertFalse(self._isTableInDb())

        testTable.create_table()
        self.assertTrue(self._isTableInDb())

    def testReadWithoutCondition(self):
        testTable = self._database.bitmask_param_table

        writtenRows = self._createBitmaskParamTableRows()
        testTable.write(writtenRows)

        parameterProvider = self.BitmaskParamTableParameterProvider()
        readRows = testTable.read(parameterProvider)
        numReadRows = 0
        for readRow in readRows:
            self._checkBitmaskParamTableRow(writtenRows[numReadRows], readRow)
            numReadRows += 1
        self.assertTrue(len(writtenRows), numReadRows)

    def testReadWithCondition(self):
        testTable = self._database.bitmask_param_table

        writtenRows = self._createBitmaskParamTableRows()
        testTable.write(writtenRows)

        parameterProvider = self.BitmaskParamTableParameterProvider()
        condition = "name='Name1'"
        readRows = testTable.read(parameterProvider, condition)
        expectedRowNum = 1
        for readRow in readRows:
            self._checkBitmaskParamTableRow(writtenRows[expectedRowNum], readRow)

    def testUpdate(self):
        testTable = self._database.bitmask_param_table

        writtenRows = self._createBitmaskParamTableRows()
        testTable.write(writtenRows)

        updateRowId = 3
        updateRow = self._createBitmaskParamTableRow(updateRowId, "UpdatedName")
        updateCondition = "id=" + str(updateRowId)
        testTable.update(updateRow, updateCondition)

        parameterProvider = self.BitmaskParamTableParameterProvider()
        readRows = testTable.read(parameterProvider, updateCondition)
        for readRow in readRows:
            self._checkBitmaskParamTableRow(updateRow, readRow)

    def _createBitmaskParamTableRows(self):
        rows = []
        for rowId in range(NUM_BITMASK_PARAM_TABLE_ROWS):
            rows.append(self._createBitmaskParamTableRow(rowId, "Name" + str(rowId)))

        return rows

    def _createBitmaskParamTableRow(self, rowId, name):
        values1 = [rowId for i in range(self._bitmaskParamTableCount1.value)]
        testBlob1 = self.api.explicit_bitmask_param.TestBlob(self._bitmaskParamTableCount1, values1)

        values2 = [rowId + 1 for i in range(self._bitmaskParamTableCount2.value)]
        testBlob2 = self.api.explicit_bitmask_param.TestBlob(self._bitmaskParamTableCount2, values2)

        values3 = [rowId + 2 for i in range(self._bitmaskParamTableCount1.value)]
        testBlob3 = self.api.explicit_bitmask_param.TestBlob(self._bitmaskParamTableCount1, values3)

        return (rowId, name, testBlob1, testBlob2, testBlob3)

    def _checkBitmaskParamTableRow(self, row1, row2):
        self.assertEqual(row1, row2)

        # check reused explicit count1 parameter
        blob1 = row2[2]
        blob3 = row2[4]
        self.assertEqual(blob1.count, blob3.count)

    def _isTableInDb(self):
        # check if database does contain table
        sqlQuery = "SELECT name FROM sqlite_master WHERE type='table' AND name='" + TABLE_NAME + "'"
        for row in self._database.connection.cursor().execute(sqlQuery):
            if len(row) == 1 and row[0] == TABLE_NAME:
                return True

        return False

TABLE_NAME = "bitmaskParamTable"
NUM_BITMASK_PARAM_TABLE_ROWS = 5
